File: 26_Triplet_Sum_in_Array/Python/Optimised.py
def has_triplet_sum(arr, target):
    n = len(arr)
    arr.sort()  # Sort the array first

    # Fix the first element and use two pointers for the remaining part
    for i in range(n - 2):
        left = i + 1
        right = n - 1

        while left < right:
            current_sum = arr[i] + arr[left] + arr[right]

            if current_sum == target:
                return True
            elif current_sum < target:
                left += 1
            else:
                right -= 1

    return False

# Sample input
arr = [1, 4, 45, 6, 10, 8]
target = 13

# Function call
print("true" if has_triplet_sum(arr, target) else "false")

"""  💡 Notes:
Time Complexity: O(n²)

Space Complexity: O(1)

Best approach for large datasets — fast and memory-efficient. """

File: 26_Triplet_Sum_in_Array/Java/Optimised.java
import java.util.Arrays;

public class Optimised {

    // Function to check if any triplet sums up to the target using sorting + two-pointer
    public static boolean hasTripletSum(int[] arr, int target) {
        int n = arr.length;

        // Sort the array to apply two-pointer technique
        Arrays.sort(arr);

        // Fix the first element and apply two-pointer on the rest
        for (int i = 0; i < n - 2; i++) {
            int left = i + 1;
            int right = n - 1;

            while (left < right) {
                int sum = arr[i] + arr[left] + arr[right];

                if (sum == target) {
                    return true;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return false;
    }

    public static void main(String[] args) {
        // Sample input
        int[] arr = {1, 4, 45, 6, 10, 8};
        int target = 13;

        // Function call
        if (hasTripletSum(arr, target)) {
            System.out.println("true");
        } else {
            System.out.println("false");
        }
    }
}

/* 💡 Notes:
Time Complexity: O(n²)

Space Complexity: O(1)

Most suitable for large arrays and optimal for real-world use. */

